hpc0.qmd

```diff
@@ -103,195 +115,513 @@ ssh <username>@arc4.leeds.ac.uk
+List all the lines in the files in `IOM-animals` directory alphabetically and find the `50th` item.
 
 ### Solution
 
-To get the `50th` line, use `head` to get the first 50 lines and `pipe it` to `tail` to get the last one:
+Use `head` to get the first 50 lines and `pipe it` to `tail` to get the last one:
 
 ```{.bash code-line-numbers=false}
-[user@host IOM-animals]$ sort *.txt | head -50 | tail -1
+[user@host IOM-animals]$ sort *.txt | head -n 50 | tail -n 1
 coot, fulica atra
 ```
+
+\
+Another option is to use the `sed` command instead of `head` and `tail`:
+```{.bash code-line-numbers=false}
+[user@host IOM-animals]$ sort *.txt | sed -n 50p
+coot, fulica atra
+```
+\
+`sed` is a program command tool for stream editing. Figuring out the `sed` options that work is not so easy! 
+
 :::
 
-## Exercise 4 {.smaller}
+
+## sed 1/2
+
+`sed` reads an input stream and produces an output stream. Internally it has a pattern space and a hold buffer. Data is read from the input stream until the next newline character and placed into the pattern space. Most sed commands operate on the data in the pattern space. Data can be exchanged between the pattern space and the hold buffer. Once sed has executed all commands, it outputs the pattern space and adds a newline characer at the end.
+
+
+## sed 2/2
+
+A very simple `sed` program:
+```
+sed 's/foo/bar/'
+```
+\
+This program replaces all instance of 'foo' with 'bar' on every line.
+
+\
+Beware of `sed` commands as they can modify a file. 
+
+
+## grep
+
+`grep` will print all lines containing or not containing a string of characters. The following will stream out all input lines containing 'foo':
+\
+```
+grep foo
+```
+\
+
+The following will stream out all input lines not containing 'foo':
+\
+```
+grep -v foo
+```
+
+
+## Exercise 6 {.smaller}
 
 ::: {.panel-tabset}
 
 ### Problem
 
-`shell-training/data/` contains 300 data files, each of which should contain 100 values. One of these files is missing some data though. . .
+`shell-training/data/` contains 300 data files, each of which should contain 100 values. One or more of these files are missing some data though...
 
 \
-Use a series of commands connected by pipes to identify the file with missing data. Hints:
+Use a series of commands connected by pipes to identify which files have missing data. Hints:
 
 - `wc -w` will tell you the number of values in a file
 - `sort -n` will sort numerically
 
 ### Solution
 
+A solution using `grep`:
 ```{.bash code-line-numbers=false}
-[user@host data]$ wc -w values* | sort -n | head -1
-99 values 44
+[user@host data]$ wc -w values* | sort -n | grep -v '100 values'
+    99 values 44
+ 29899 total
 ```
 
-- `wc -w` tells you how many values are in each file
-- `sort` this output to put the files in order of the number of values
-- use `head` to select the first line, which contains the name of the file with the smallest number of values
+\
```

Comment:
Let's not get into awk.  It's just going to terrify people.

Reply:
I have left in a couple of awk examples and a brief bit about what awk is. We can completely skip it, but I think it is good to have in. There is no need for them to use it for the exercises and it is not part of the main slides or commands that are introduced.